File: app/src/main/java/com/stormbreakerstudios/android/challenge/OnboardingActivity.java
package com.stormbreakerstudios.android.challenge;

import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;

import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;

import com.stormbreakerstudios.android.challenge.databinding.ActivityOnboardingBinding;
import com.stormbreakerstudios.android.challenge.util.DatePickerFragment;
import com.stormbreakerstudios.android.challenge.util.TimePickerFragment;


public class OnboardingActivity extends BaseActivity implements DatePickerFragment.DatePickerListener,
        TimePickerFragment.TimePickerListener, AdapterView.OnItemSelectedListener {

    private ActivityOnboardingBinding viewBinding = null;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        viewBinding = ActivityOnboardingBinding.inflate(getLayoutInflater());
        setContentView(viewBinding.getRoot());
        setTransparentStatusBarForKitKatAndAbove();
        setMarginForSaveButton();
    }

    private void setMarginForSaveButton() {
        int resourceId = getResources().getIdentifier("navigation_bar_height", "dimen", "android");
        int navbarHeight = getResources().getDimensionPixelSize(resourceId);

        ConstraintLayout.LayoutParams params = (ConstraintLayout.LayoutParams) viewBinding.btSave.getLayoutParams();
        params.setMargins(
                params.leftMargin,
                params.topMargin,
                params.rightMargin,
                params.bottomMargin + navbarHeight
        );
        viewBinding.btSave.setLayoutParams(params);
    }

    public void showDatePickerDialog(View view) {
        DatePickerFragment newFragment = new DatePickerFragment(this);
        newFragment.show(getSupportFragmentManager(), "datePicker");
    }

    public void showTimePickerDialog(View view) {
        TimePickerFragment newFragment = new TimePickerFragment(this);
        newFragment.show(getSupportFragmentManager(), "timePicker");
    }

    public void showTimeZoneSpinnerDialog(View view) {
        Spinner spinner = viewBinding.spTimeZone;
        ArrayAdapter<CharSequence> arrayAdapter = ArrayAdapter.createFromResource(
                this, R.array.timezone_array, android.R.layout.simple_spinner_item
        );
        arrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spinner.setAdapter(arrayAdapter);
        spinner.setOnItemSelectedListener(this);
        spinner.performClick();
    }

    public void onDatePicked(int year, int month, int day) {
        viewBinding.etBirth.setText(String.format("%02d/%02d/%s", month, day, year));
    }

    public void onTimePicked(int hourOfDay, int minute) {
        viewBinding.etMeditationTime.setText(String.format("%02d:%02d", hourOfDay, minute));
    }

    public void onNothingSelected(AdapterView<?> parent) {
    }

    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
        viewBinding.etTimeZone.setText(parent.getItemAtPosition(position).toString());
    }

    private boolean isFormValid() {
        return !viewBinding.etBirth.getText().toString().isEmpty()
                && !viewBinding.etMeditationTime.getText().toString().isEmpty()
                && !viewBinding.etTimeZone.getText().toString().isEmpty();
    }

    public void onBoardSaved(View view) {
        if (isFormValid()) {
            navigator.navigateToMainScreen(this);
            finish();
        } else
            toastLong("There are missing fields to fill.");
    }
}

File: app/src/main/java/com/stormbreakerstudios/android/challenge/OnboardingActivity.kt
package com.stormbreakerstudios.android.challenge

import android.os.Bundle
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import androidx.constraintlayout.widget.ConstraintLayout
import com.stormbreakerstudios.android.challenge.databinding.ActivityOnboardingBinding
import com.stormbreakerstudios.android.challenge.util.DatePickerFragment
import com.stormbreakerstudios.android.challenge.util.TimePickerFragment


class OnboardingActivity : BaseActivity(), DatePickerFragment.DatePickerListener,
    TimePickerFragment.TimePickerListener, AdapterView.OnItemSelectedListener {

    private val TAG = OnboardingActivity::class.java.simpleName

    private lateinit var viewBinding: ActivityOnboardingBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewBinding = ActivityOnboardingBinding.inflate(layoutInflater)
        setContentView(viewBinding.root)
        setTransparentStatusBarForKitKatAndAbove()
        setMarginForSaveButton()
    }

    private fun setMarginForSaveButton() {
        val resourceId = resources.getIdentifier("navigation_bar_height", "dimen", "android")
        val navbarHeight = resources.getDimensionPixelSize(resourceId)

        val params = viewBinding.btSave.layoutParams as ConstraintLayout.LayoutParams
        params.setMargins(
            params.leftMargin,
            params.topMargin,
            params.rightMargin,
            params.bottomMargin + navbarHeight
        )
        viewBinding.btSave.layoutParams = params
    }

    fun showDatePickerDialog(view: View) {
        val newFragment = DatePickerFragment(this)
        newFragment.show(supportFragmentManager, "datePicker")
    }

    fun showTimePickerDialog(view: View) {
        val newFragment = TimePickerFragment(this)
        newFragment.show(supportFragmentManager, "timePicker")
    }

    fun showTimeZoneSpinnerDialog(view: View) {
        val spinner = viewBinding.spTimeZone
        ArrayAdapter.createFromResource(
            this, R.array.timezone_array, android.R.layout.simple_spinner_item
        ).also { adapter ->
            adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
            spinner.adapter = adapter
            spinner.onItemSelectedListener = this
            spinner.performClick()
        }
    }

    override fun onDatePicked(year: Int, month: Int, day: Int) {
        viewBinding.etBirth.text = String.format("%02d/%02d/%s", month, day, year)
    }

    override fun onTimePicked(hourOfDay: Int, minute: Int) {
        viewBinding.etMeditationTime.text = String.format("%02d:%02d", hourOfDay, minute)
    }

    override fun onNothingSelected(parent: AdapterView<*>?) {
    }

    override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
        viewBinding.etTimeZone.text = parent?.getItemAtPosition(position).toString()
    }

    private fun isFormValid(): Boolean {
        return viewBinding.etBirth.text.isNotEmpty()
                && viewBinding.etMeditationTime.text.isNotEmpty()
                && viewBinding.etTimeZone.text.isNotEmpty()
    }

    fun onBoardSaved(view: View) {
        if (isFormValid())
            finish()
        else
            toastLong("There are missing fields to fill.")
    }
}